File: composeApp/src/androidMain/kotlin/com/arviejhay/secretboard/ViewModel/BoardViewModel.kt
package com.arviejhay.secretboard.ViewModel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.arviejhay.secretboard.data.model.Post
import com.arviejhay.secretboard.data.usecase.GetPostsForDataUseCase
import com.arviejhay.secretboard.data.usecase.SubmitPostUseCase
import com.arviejhay.secretboard.domain.api.Result.GetPostResult
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.launch

class BoardViewModel(
    private val getPosts: GetPostsForDataUseCase,
    private val submitPost: SubmitPostUseCase
) : ViewModel() {

    val posts = MutableStateFlow<List<Post>>(emptyList())
    val isLoading = MutableStateFlow(false)
    val hasAnError = MutableStateFlow(false)

    fun loadPosts(date: String) {
        hasAnError.value = false
        isLoading.value = true

        viewModelScope.launch {
            val result = getPosts(date)
            when (result) {
                is GetPostResult.Success -> posts.value = result.posts
                is GetPostResult.Failure -> posts.value = emptyList()
            }
            hasAnError.value = result is GetPostResult.Failure
            isLoading.value = false
        }
    }

    fun sendPost(name: String?, title: String?, message: String) {
        viewModelScope.launch {
            submitPost(name, title, message)
            loadPosts("2023-10-01")
        }
    }
}

File: iosApp/iosApp/ViewModel/BoardViewModel.swift
//
//  BoardViewModel.swift
//  iosApp
//
//  Created by Arviejhay Alejandro on 8/6/25.
//

import Shared

class BoardViewModel: ObservableObject {
    var getPostsUseCase: GetPostsForDataUseCase
    var submitPostsUseCase: SubmitPostUseCase
    
    @Published var posts: [Post] = []
    
    init(getPostsUseCase: GetPostsForDataUseCase, submitPostsUseCase: SubmitPostUseCase) {
        self.getPostsUseCase = getPostsUseCase
        self.submitPostsUseCase = submitPostsUseCase
    }
    
    func loadPosts() async {
        do {
            let result = try await getPostsUseCase.invoke(date: "2025-01-01")
            switch result {
            case let success as GetPostResult.Success:
                self.posts = success.posts
            case let failure as GetPostResult.Failure:
                print(failure.exception)
            default:
                break
            }
        } catch {
            // Handle error (e.g., log or set an error state)
            print("Failed to load posts: \(error)")
        }
    }
}
